### PR TITLE
Multithreading refactor for ObjectManager.

### PR DESCRIPTION
**Pull request by elibol**
This PR refactors the ObjectManager to send/receive messages and create connections on the main thread. It also removes the ObjectManager queuing mechanism and instead relies on asio's internal queuing.

### CODE REVIEW

**Review by pcmoritz — APPROVED** [review]
LGTM, I went through it with Melih, @stephanie-wang: Can you also have a look?

File: src/ray/object_manager/object_manager.cc [hunk 1]
```diff
@@ -325,14 +256,12 @@ ray::Status ObjectManager::SendObjectHeaders(const ObjectID &object_id,
...
 
-  if (!chunk_status.second.ok()) {
-    // This is the first thread to invoke GetChunk => Get failed on the
-    // plasma client.
-    // No reference is acquired for this chunk, so no need to release the chunk.
-    // TODO(hme): Retry send here? If so, store RemoteConnectionInfo in SenderConnection.
-    RAY_CHECK_OK(TransferCompleted(TransferQueue::TransferType::SEND));
-    return chunk_status.second;
-  }
+  // If status is not okay, and this is the first chunk to be gotten for this object,
```
Comment (stephanie-wang) [inline src/ray/object_manager/object_manager.cc]: The comment about "this is the first chunk to be gotten" isn't correct anymore, right? Can we remove that?

**Review by stephanie-wang — APPROVED** [review]
Looks good! Can you update the comment that I pointed out before merging?

File: src/ray/object_manager/object_manager.cc (hunk 1 above)
Comment (elibol) [reply src/ray/object_manager/object_manager.cc]: The comment seems to be correct to me. Can you explain why it's no longer correct?

Comment (stephanie-wang) [reply src/ray/object_manager/object_manager.cc]: Shouldn't it be "If status is not okay, then return immediately..."? I don't think it matters which chunk of the object it is.

**Pull request merged**